Add GitLab integration for scaffolder

muffix opened the pull request:
## Hey, I just made a Pull Request!

<!-- Please describe what you added, and add a screenshot if possible.
     That makes it easier to understand the change so we can :shipit: faster. -->

This adds a GitLab integration for the scaffolder backend.

We're introduceing a preparer and a publisher for GitLab so that we can
read templates from GitLab and publish them to a configured GitLab
instance. The two instances don't need to be the same. For instance,
templates could be public on gitlab.com, but the created repos will live
in a hosted GitLab somewhere else.

The publisher gets its own config object in `app-config.yaml` where the
target instance and token can be specified.

The service catalogue defines both `gitlab` and `gitlab/api` as
processors. They are both handled by the same preparer.

Closes #2372

#### :heavy_check_mark: Checklist

<!--- Put an `x` in all the boxes that apply: -->

- [x] All tests are passing `yarn test`
- [ ] Screenshots attached (for UI changes)
- [x] Relevant documentation updated
- [x] Prettier run on changed files
- [x] Tests added for new functionality
- [x] Regression tests added for bug fixes

muffix commented on plugins/scaffolder-backend/package.json:
```diff
@@ -34,7 +36,7 @@
     "express-promise-router": "^3.0.3",
     "fs-extra": "^9.0.0",
-    "git-url-parse": "^11.1.2",
+    "git-url-parse": "^11.2.0",
```
This version bump is required because the library didn't handle GitLab URLs with scope correctly (see IonicaBizau/git-url-parse#107).

muffix commented on plugins/scaffolder-backend/src/scaffolder/stages/prepare/file.ts:
```diff
@@ -17,7 +17,7 @@ import fs from 'fs-extra';
 import os from 'os';
 import { TemplateEntityV1alpha1 } from '@backstage/catalog-model';
-import { parseLocationAnnotation } from './helpers';
+import { parseLocationAnnotation } from '../helpers';
```
Helper functions are now used by both preparers and publishers, so they moved a directory up.

muffix commented on plugins/scaffolder-backend/src/scaffolder/stages/prepare/gitlab.ts:
```diff
@@ -0,0 +1,73 @@
+
+  constructor(config: Config) {
+    this.privateToken =
+      config.getOptionalString('catalog.processors.gitlabApi.privateToken') ??
```
This is taking the private GitLab token from the catalogue's config. Technically, we're crossing packages here and we might want to configure the preparer token explicitly. But since we're only going to use this token to clone the repo which the catalogue already knows how to do (and is allowed to), this might be okay. What do you think?

benjdlambert replied:
Hmm, this is interesting. 

I'd suggest two different tokens for now, and place it under the scaffolder config, and accept the duplication for now.

We're planning on moving to using a GithubApp for authentication for the backend plugins for Github, and I think we will be doing the same for GitLab too once we have something that works

muffix commented on plugins/scaffolder-backend/src/scaffolder/stages/publish/gitlab.ts:
```diff
@@ -0,0 +1,90 @@
+  }
+
+  private async pushToRemote(directory: string, remote: string): Promise<void> {
+    const repo = await Repository.init(directory, 0);
```
This is a lot of repetition. We could refactor initialising and crating the repo into an abstract base class to avoid that.

muffix commented on plugins/scaffolder-backend/src/scaffolder/stages/publish/publishers.ts:
```diff
@@ -0,0 +1,39 @@
+    this.publisherMap.set(protocol, publisher);
+  }
+
+  get(template: TemplateEntityV1alpha1): PublisherBase {
```
I'm not sure if it makes sense to derive the publisher from the entity. A template can absolutely be stored in, e.g., public GitHub and the newly created repo should be pushed to a private GitLab instance. It'd probably be best to let the user choose if there's more than one publisher configured, but for now, I think it's probably a fair assumption that in most cases the target repo and the template will live in the same GitHub/GitLab instance.

benjdlambert replied:
That is 100% the end goal. We ideally want to have a list of publishers to choose from, but we just haven't gotten to that stage yet.

Ideally we would be able to do `/v1/scaffolder/publishers` and display this list of publishers in the frontend when choosing the stages.

I will create a ticket for this.

muffix reviewed (commented):
Added some annotations/thoughts which hopefully make it easier to review.

benjdlambert commented:
I will take a look a this later on today. Thanks for the contribution!

benjdlambert commented:
Sorry the core teams been pretty busy with some team meetings for the last few days, we're back to normal work tomorrow, so I will pick this up then! Hope that's OK, and really sorry for the delay.

muffix commented:
Of course, don't worry. Thanks for taking the time.

mfrinnstrom commented:
Hi,

I'm interested to see where this goes, both when it comes to the config but also the base classes, since I will start on the same work but for Azure DevOps. 

My plan is also to work on the templater parts since I will need to run the containers on ECS (using Fargate) instead of a local Docker daemon.

benjdlambert commented on app-config.yaml:
```diff
@@ -36,6 +36,14 @@ techdocs:
   requestUrl: http://localhost:7000/techdocs/docs
 
+scaffolder:
+  gitlab:
```
See comment https://github.com/spotify/backstage/pull/2426#discussion_r489405681. Think here is fine for now, but the keys will move away eventually.

benjdlambert commented on plugins/scaffolder-backend/src/scaffolder/stages/publish/publishers.ts:
```diff
@@ -0,0 +1,39 @@
+
+  get(template: TemplateEntityV1alpha1): PublisherBase {
+    const { protocol } = parseLocationAnnotation(template);
+    const publisher = this.publisherMap.get(protocol);
```
I have a small concern here that when using the `file:` location in the catalog, this will break as there is no `file:` publisher registered. Maybe we can fallback to the `github` one for now. And throw an error if that doesn't exist.

It's a little hard without doing what's mentioned here:  https://github.com/spotify/backstage/pull/2426/files#r489404507

benjdlambert replied:
It would also be good to document this new behaviour in https://github.com/spotify/backstage/blob/master/docs/features/software-templates/installation.md and documention the above behaviour

muffix replied:
I believe that's already covered by us registering the GitHub publisher for the `file` protocol: https://github.com/spotify/backstage/pull/2426/files#diff-b7b27027cea69fa63c131d6e290c9156R59

benjdlambert reviewed (changes requested):
This is awesome work. Just a few comments I left could do with replies and the `file`preparers stuff

mfrinnstrom commented:
Another comment on the config part. I would like it under a separate scaffolding section since then there is a natural place to add other things like a base path that would be useful if you have another container orchestrator for the templater parts and you mount networked storage between the containers.

muffix commented:
Thanks, I rebased the branch on the latest master and updated a test. I also moved the documentation to the installation.md as requested.

Happy to squash the commits and rebase again once you're happy @benjdlambert.

muffix commented:
Resolved the conflicts that merging #2429 caused. #2429 also decided for us which style of config file to use, which makes it easier for us.

benjdlambert commented:
Would like a quick review from @Rugvip too, just to see if theres anything i've missed, and of course theres @spotify/techdocs-core

benjdlambert commented:
Also looks like there's some build failures too @muffix

OrkoHunter commented on docs/features/software-templates/installation.md:
```diff
@@ -189,17 +211,31 @@ docs on creating private GitHub access tokens is available
+      baseUrl: https://gitlab.com
+      token:
+        $secret:
+          env: SCAFFOLDER_GITLAB_PRIVATE_TOKEN
```
Does it make sense to add consistency in the name of these tokens for GitHub/GitLab?

`SCAFFOLDER_GITLAB_PRIVATE_TOKEN` can be `GITLAB_PRIVATE_TOKEN` to align with `GITHUB_ACCESS_TOKEN`

OrkoHunter replied:
Edit: `GITLAB_ACCESS_TOKEN` is more aligned.

muffix replied:
I was thinking that there could be a naming clash if we're using different access tokens for different accounts/plugins. That's something we can worry about later, though. Changed it to `GITLAB_ACCESS_TOKEN`. 👍

OrkoHunter commented on plugins/scaffolder-backend/src/scaffolder/stages/publish/gitlab.ts:
```diff
@@ -0,0 +1,90 @@
+    const oid = await index.writeTree();
+    await repo.createCommit(
+      'HEAD',
+      Signature.now('Scaffolder', 'scaffolder@backstage.io'),
+      Signature.now('Scaffolder', 'scaffolder@backstage.io'),
```
@spotify/backstage-core Does this email ID really exists? It makes sense to have a public "backstage-service-account" or "spotify-oss-service-account". (Excuse me if there is already one 🙇 )

The account could also be tied up with any GitHub app we might create and provide for authentication in backstage.

benjdlambert replied:
So it doesn't exist yet, but it will be tied with the GIthub App and GitLab app when that is completed

muffix replied:
Yeah, this is exactly the same behaviour as for GitHub

OrkoHunter reviewed (approved):
Left couple comments. Approving this to unblock @spotify/backstage-core on merging this PR. 🙇

muffix commented:
Thanks @OrkoHunter, @benjdlambert. I must have borked the YAML of the config when resolving the merge conflicts. Fixed that and rebased again. This should make the build green again.

timja commented:
for the crash on master if you don't define the gitlab token: https://github.com/spotify/backstage/pull/2500
